### DISCUSSION TITLE
SES shim depends on console having certain methods

### DISCUSSION THREAD

**Discussion by p-bakker** [endojs/endo#2643] [Q&A]
For example .groupEnd or .error

Some environments don't have all those methods, for example QuickJS-NG, which only has .log

Is SES supported only on environments that have a full console.log impl. or does SES incorrectly depend on certain methods being available on console?

**Comment by mhofman** [comment]
Yeah the SES shim currently relies on a fully defined `console` object. That's arguably a mis-layering since console is not part of the language. 

There are 2 ways the SES shim interacts with the console today:
- it uses it to provide lockdown diagnostics, and
- it tames it to reveal causal error information when logged.

While we have plans to eject the console "taming" into a trusted shim with SES just leaving a power in the start compartment to reveal error information, we should also fix the reliance of the SES shim onto the console for diagnostics, and instead make a best effort, possibly falling back to other known print mechanisms.

In the meantime you might be interested in https://github.com/Agoric/agoric-sdk/blob/master/packages/xsnap-lockdown/lib/console-shim.js which we use on XS that similarly lacks a proper console. I'm also wondering how @kriskowal is addressing this in his more direct support of XS in https://github.com/endojs/endo/pull/2471.

**Comment by kriskowal** [comment]
We recently changed our internal usage of `console` to go through a `reporter` abstraction and reporter implementations chosen based on a `reporting` option on `lockdown`. It now detects whether the console implements `groupCollapsed` or `groupEnd` are implemented and fall back to an emulation of those if they are absent, but does not fall back to `log` if `warn` and `error` are absent. It does fall back to `print` if `console` is absent. We can refine this. Issues are welcome. Specific engines to support would be welcome.

**Reply by p-bakker** [reply]
I assume those changes are in 1.10.0? Am on that version and when that version goes the console.xxxx route, it assumes that console.error is available (code crashes otherwise). And it assumes that if console.group exists, that console.groupEnd also exist.

Both assumptions don't hold true in QuickJs(-NG).

Wondering whether the logic that uses console.group should check for all required methods before starting to use the group functionality

And whether the code should fallback to just using console.log if console.error doesn't exist

**Reply by p-bakker** [reply]
Looks like something for done for console.groupEnd: https://github.com/endojs/endo/commit/fd70235cb697b8c75d4e61d6b017f11d708798ad
